### PR TITLE
REPL shell mode for Windows

### PR DESCRIPTION
**Pull request by musm**
Previous Attempts: https://github.com/JuliaLang/julia/pull/31490 https://github.com/JuliaLang/julia/pull/23703 https://github.com/JuliaLang/julia/pull/21294 https://github.com/JuliaLang/julia/pull/20776 https://github.com/JuliaLang/julia/pull/11478 https://github.com/JuliaLang/julia/pull/33427 https://github.com/JuliaLang/julia/pull/11478

Some of these approaches tried to implement special Windows rule in the `shell_escape` mechanism, I think some also tried to pass 'raw' strings to the shell, all with varying degrees of success.

Fixes https://github.com/JuliaLang/julia/issues/6316 https://github.com/JuliaLang/julia/issues/23597

**References**

Powershell:

- https://www.red-gate.com/simple-talk/wp-content/uploads/imported/2264-PS%20Quoting%20Wallchart_1_0_1.pdf
- https://stackoverflow.com/questions/44981177/run-powershell-custom-function-in-new-window/44986341#44986341
- https://ss64.com/ps/syntax-esc.html

CMD:

- https://www.robvanderwoude.com/escapechars.php
- https://ss64.com/nt/syntax-esc.html

### CODE REVIEW

File: base/Base.jl [hunk 1]
```diff
@@ -174,8 +174,8 @@ include("iobuffer.jl")
 include("intfuncs.jl")
 include("strings/strings.jl")
 include("parse.jl")
-include("shell.jl")
 include("regex.jl")
+include("shell.jl")
```
Comment (musm) [inline base/Base.jl]: This change is needed since we use substitution string in the `shell.jl` file.

File: base/client.jl [hunk 2]
```diff
@@ -31,8 +31,15 @@ stackframe_lineinfo_color() = repl_color("JULIA_STACKFRAME_LINEINFO_COLOR", :bol
 stackframe_function_color() = repl_color("JULIA_STACKFRAME_FUNCTION_COLOR", :bold)
 
 function repl_cmd(cmd, out)
-    shell = shell_split(get(ENV, "JULIA_SHELL", get(ENV, "SHELL", "/bin/sh")))
+    shell = get(ENV, "JULIA_SHELL", nothing)
```
Comment (musm) [inline base/client.jl]: Should we cache this value in a reference? So that we don't have to look it up each time the  Julia REPL is entered?

File: base/client.jl [hunk 3]
```diff
@@ -66,15 +73,28 @@ function repl_cmd(cmd, out)
...
-        @static if !Sys.iswindows()
+        local command::Cmd
+        if Sys.iswindows()
+            if shell_name == ""
+                command = cmd
+            elseif shell_name == "cmd"
+                command = Cmd(`$shell /c $(shell_escape_CMDly(shell_escape_winsomely(cmd)))`, windows_verbatim=true)
+            elseif shell_name in ("powershell", "pwsh")
+                command = Cmd(`$shell -Command $(shell_escape_PWSHly(shell_escape_winsomely(cmd)))`, windows_verbatim=true)
+            elseif shell_name == "busybox"
```
Comment (musm) [inline base/client.jl]: Is this really necessary ? We don't even ship with busy box anymore...

File: base/shell.jl [hunk 4]
```diff
@@ -253,3 +253,129 @@ julia> Base.shell_escape_posixly("echo", "this", "&&", "that")
...
+    to_print = "All for 1 & 1 for all!"
+    run(Cmd(Cmd(["cmd /c \"break | echo \$(shell_escape_CMDly(shell_escape_CMDly(to_print)))"]), windows_verbatim=true))
+    ```
+"""
+shell_escape_CMDly(arg::AbstractString) = sprint(print_shell_escaped_CMDly, arg)
+
+function print_shell_escaped_BATCHly(io::IO, arg::AbstractString)
+    # see https://www.robvanderwoude.com/variableexpansion.php
+    # the rule for ! is a bit esoteric, but doesn't hurt to add it
+    any(in("\r\n"), arg) && throw("Encountered unsupported character by CMD")
```
Comment (musm) [inline base/shell.jl]: make this an `ArgumentError` error?

File: base/shell.jl [hunk 5]
```diff
@@ -253,3 +253,129 @@ julia> Base.shell_escape_posixly("echo", "this", "&&", "that")
...
+    ```
+    to_print = "All for 1 & 1 for all!"
+    run(Cmd(Cmd(["cmd /c \"break | echo \$(shell_escape_CMDly(shell_escape_CMDly(to_print)))"]), windows_verbatim=true))
+    ```
+"""
+shell_escape_CMDly(arg::AbstractString) = sprint(print_shell_escaped_CMDly, arg)
+
+function print_shell_escaped_BATCHly(io::IO, arg::AbstractString)
+    # see https://www.robvanderwoude.com/variableexpansion.php
+    # the rule for ! is a bit esoteric, but doesn't hurt to add it
```
Comment (vtjnash) [inline base/shell.jl]: I think that's just because folks are bad at it. It seems clearly part of the list to me. Some sources list `(` and `)`, but—as explained in https://stackoverflow.com/questions/4094699/how-does-the-windows-command-interpreter-cmd-exe-parse-scripts—they are not needed. Nor is `@`, as is claimed by this otherwise fairly good source: http://daviddeley.com/autohotkey/parameters/parameters.htm#WINCMDRULES

File: base/shell.jl [hunk 6]
```diff
@@ -253,3 +253,129 @@ julia> Base.shell_escape_posixly("echo", "this", "&&", "that")
...
+    arg = replace(arg, r"[()!^\"<>&|]" => s"^\0")
+    arg = replace(arg, "%" => "%%")
+    print(io, arg)
+end
+
+"""
+     shell_escape_BATCHly(arg::AbstractString)::String
+
+Like [`shell_escape_CMDly`](@ref), but appropriate for use inside batch (*.bat) files for use
+with `CMD.exe`.
```
Comment (vtjnash) [inline base/shell.jl]: actually, this might be confusing to folks, so perhaps we should not have this function. I think I'd naively expect this to be the function to quote arguments to a batch script, while this actually quotes them to be values inside a batch script (and it seems much less likely that someone is going to emit `*.bat` files and then run them)

File: base/client.jl (hunk 3 above)
Comment (musm) [reply base/client.jl]: arg .. looks like it is needed for testing on windows...  maybe we should look into WSL  for testing on windows instead of downloading busybox?

Comment (musm) [reply base/client.jl]: Still something about this rubs me the wrong way. We should still try to remove this and try to work around them in the tests. Thoughts welcome

File: base/shell.jl [hunk 7]
```diff
@@ -312,3 +312,51 @@ julia> println(shell_escaped_winsomely("A B\\", "C"))
...
+    ```
+    to_print = "All for 1 & 1 for all!"
+    run(Cmd(Cmd(["cmd /c \"break | echo \$(shell_escape_CMDly(shell_escape_CMDly(to_print)))"]), windows_verbatim=true))
+    ```
+"""
+shell_escape_CMDly(arg::AbstractString) = sprint(print_shell_escaped_CMDly, arg)
+
+function print_shell_escaped_PWSHly(io::IO, arg::AbstractString)
+    arg = replace(arg, r"[\$#;]" => s"`\0")
+    arg = replace(arg, r"[\"|\\\\(?=\\\\*(\"|$))]" => s"\\\0")
```
Comment (musm) [inline base/shell.jl]: https://stackoverflow.com/questions/44981177/run-powershell-custom-function-in-new-window/44986341#44986341

File: base/shell.jl [hunk 8]
```diff
@@ -312,3 +312,51 @@ julia> println(shell_escaped_winsomely("A B\\", "C"))
...
+    (indicating a command pipeline) must have `shell_escape_CMDly` applied twice. For example:
+    ```
+    to_print = "All for 1 & 1 for all!"
+    run(Cmd(Cmd(["cmd /c \"break | echo \$(shell_escape_CMDly(shell_escape_CMDly(to_print)))"]), windows_verbatim=true))
+    ```
+"""
+shell_escape_CMDly(arg::AbstractString) = sprint(print_shell_escaped_CMDly, arg)
+
+function print_shell_escaped_PWSHly(io::IO, arg::AbstractString)
+    arg = replace(arg, r"[\$#;]" => s"`\0")
```
Comment (musm) [inline base/shell.jl]: `\$` prevents  powershell variables, `;` prevents  executing two commands, `#` escapes comment character

File: base/shell.jl (hunk 7 above)
Comment (vtjnash) [reply base/shell.jl]: That left me even more confused. It ends with but it’s not this simple, and the rules have changed over time. 

One thing I’m fairly confident of is that this pwsh function needs us to add \r and \n to the winsomely whitespace list to pass through correctly here (no effect if it doesn’t go to pwsh, so might as well always do it there).

Comment (musm) [reply base/shell.jl]: Hmm, yeah I'm confused now as well too, I don't think we need it.

> One thing I’m fairly confident of is that this pwsh function needs us to add \r and \n to the winsomely whitespace list to pass through correctly here (no effect if it doesn’t go to pwsh, so might as well always do it there).

Why's that?

Comment (musm) [reply base/shell.jl]: Initially, I thought we needed powershell to work for this example
```
# works with the added regex
shell> gci C:\\Program\ Files
```
why `\ `? well hitting tab-complete with `shell> gci C:\\Program` spits out `C:\\Program\ Files`, so I figured this should work. The regex fixes this.

BUT the escape for white-spaces is ``` `    ``` , so this doesn't seem necessary. 

Is it at all a odd we have to julia escape backslashes in `shell>` mode or is that normal.

**Comment by musm** [comment]
There was a bug here when someone passed `ENV["JULIA_SHELL"] = ""`, which is fixed here by also defaulting this case to either `sh` or `cmd`.

**Comment by musm** [comment]
@vtjnash how does this PR look now?

File: base/shell.jl (hunk 7 above)
Comment (musm) [reply base/shell.jl]: FYI updated my comments.

Comment (musm) [reply base/shell.jl]: @vtjnash  Please see

> The string needs escaping, however, in order to be passed to the new Powershell process unmodified.
> " instances embedded in the string are stripped, unless they are either represented as \" or tripled (""").
> (\" rather than the usual `" is needed to escape double quotes in this case, because PowerShell expects \" when passing a command to the powershell.exe executable.)


I feel confident that we now have a correct escaping solution to powershell and also for CMD

**Comment by musm** [comment]
Ok so I feel like I have finally converged to a solution here, and this should be ready for final review.

**Comment by musm** [comment]
My only question to @vtjnash  is why do you feel like we also need to  quote `\r` and `\n` in the `print_shell_escaped_winsomely` function?

**Comment by musm** [comment]
Should we default to `CMD` on Windows  OR powershell with `-noProfile` ?

File: base/client.jl [hunk 9]
```diff
@@ -66,15 +68,27 @@ function repl_cmd(cmd, out)
...
         println(out, pwd())
     else
-        @static if !Sys.iswindows()
+        local command
+        if Sys.iswindows()
+            if shell_name == "cmd"
+                command = Cmd(`$shell /c $(shell_escape_CMDly(shell_escape_winsomely(cmd)))`, windows_verbatim=true)
+            elseif shell_name in ("powershell", "pwsh")
+                # enclose in double quotes to enusre we prevent whitespace normalization
+                command = Cmd(`$shell -Command "$(shell_escape_PWSHly(shell_escape_winsomely(cmd)))"`, windows_verbatim=true)
```
[the remark this answers is not in the document]
Comment (vtjnash) [reply base/client.jl]: Although this is only correct so long as `powershell` is a no-op though. If something is getting intercepted by powershell, we need to use completely different rules. Something like:

```suggestion
CommandType = read(`powershell -Command "Get-Command -- $(shell_escape_PWSH_cmdlet_ly(cmd.exec[1])) | Select-Object -ExpandProperty CommandType"`, String)
# TODO: while CommandType == "Alias"; CommandType = ...; end
if CommandType == "Application"
    command = Cmd(`$shell -Command "& $(shell_escape_PWSHly(shell_escape_winsomely(cmd)))"`)
else # handle Function and Cmdlet # TODO: what is the proper handling for the other types (ExternalScript, Script, Workflow, Configuration, and Filter)
    command = Cmd(`$shell -Command "& $(shell_escape_PWSH_cmdlet_ly(cmd))"`)
end
```

File: base/shell.jl (hunk 8 above)
Comment (vtjnash) [reply base/shell.jl]: hm, I think it additionally also needs ```"\r" => "`r", "\n" => "`n", "\t" => "`t"```

File: base/shell.jl (hunk 7 above)
Comment (vtjnash) [reply base/shell.jl]: I wasn't sure what your source was for `gci`, so I wrote a simple one to be able to inspect what reaches the target application. Save as `gci.c`, compile with `cc gci.c`, then run as `a.exe`.
```
#include <windows.h>
#include <stdio.h>
int main (int argc, char **argv) {
  printf("%s\n", GetCommandLineA());
  for (int i = 0; i < argc; i++) { printf("%d: %s\n", i, argv[i]); }
}
```

**Comment by musm** [comment]
We should probably also allow `pwsh` on Linux and MacOS since  Powershell Core is cross-platform

File: base/client.jl [hunk 10]
```diff
@@ -66,19 +68,52 @@ function repl_cmd(cmd, out)
...
+                command = `$shell sh -c $(shell_escape_posixly(cmd))`
+            else
+                command = `$shell $cmd`
+            end
+        else
             if shell_name == "fish"
                 shell_escape_cmd = "begin; $(shell_escape_posixly(cmd)); and true; end"
+                command = `$shell -c $shell_escape_cmd`
+            elseif shell_name == "pwsh"
+                command = _powershell_execute(shell, cmd)
```
Comment (stevengj) [inline base/client.jl]: Doesn't this do the wrong escaping on Unix?

Comment (stevengj) [reply base/client.jl]: (I'm not sure it's worth the effort to support `pwsh` on Unix.   This is only used for interactive mode, after all.)

Comment (vtjnash) [reply base/client.jl]: Powershell's handling of arguments is known (https://github.com/PowerShell/PowerShell/issues/1995) to be buggy, so can't reliably be used for launching programs anyways.

Comment (stevengj) [reply base/client.jl]: @vtjnash, that's argument handling on Windows.  Presumably on Unix systems `pwsh` just uses `*argv`, no?

Comment (vtjnash) [reply base/client.jl]: Sadly, no (e.g. https://github.com/PowerShell/PowerShell/issues/10440)

> no CommandLineToArgvW layer

Sadly, it is currently present (although both broken and undocumented—and eventually supposed to be unnecessary)

**Comment by musm** [comment]
should we try to get this in, in some form before feature freze?

**Comment by StefanKarpinski** [comment]
It is ready for re-review by @vtjnash and/or @stevengj then?

File: base/client.jl (hunk 3 above)
Comment (vtjnash) [reply base/client.jl]: I think we should make this the `else` block

File: base/shell.jl [hunk 11]
```diff
@@ -312,3 +312,69 @@ julia> println(shell_escaped_winsomely("A B\\", "C"))
...
+
+!important
+    Due to a peculiar behavior of the CMD, each command after a literal `|` character
+    (indicating a command pipeline) must have `shell_escape_CMDly` applied twice. For example:
+    ```
+    to_print = "All for 1 & 1 for all!"
+    run(Cmd(Cmd(["cmd /c \"break | echo \$(shell_escape_CMDly(shell_escape_CMDly(to_print)))"]), windows_verbatim=true))
+    ```
+"""
+shell_escape_CMDly(arg::AbstractString) = sprint(print_shell_escaped_CMDly, arg)
```
Comment (vtjnash) [inline base/shell.jl]: This appears to need tests. Also, can you split this aspect into a separate commit?

**Review by vtjnash — COMMENTED** [review]
Can you remove all of the `powershell`-related content? The rest I think seems ready to go. We should add the powershell portions in a pair of followup PRs (one to add the `shell_escape_PWSHly` layer and one to use it in the REPL)

File: base/client.jl (hunk 3 above)
Comment (musm) [reply base/client.jl]: What exactly do you mean by "making this the else block"? 
Do you mean :

```julia
           if shell_name == "cmd"
                command = _CMD_execute(shell, cmd)
            elseif shell_name in ("powershell", "pwsh")
                command = _powershell_execute(shell, cmd)
            else
                  command  =  `$shell sh -c $(shell_escape_posixly(cmd))`  
```
Is the assumption here that if you don't want `cmd` or `powershell`  then you probably want it to default to use a  linux-based shell running in a windows environment?

**Comment by musm** [comment]
> Can you remove all of the powershell-related content? The rest I think seems ready to go. We should add the powershell portions in a pair of followup PRs (one to add the shell_escape_PWSHly layer and one to use it in the REPL)

I think I'll just open a seperate PR with the suggestions

File: base/shell.jl [hunk 12]
```diff
@@ -312,3 +312,69 @@ julia> println(shell_escaped_winsomely("A B\\", "C"))
 """
 shell_escape_winsomely(args::AbstractString...) =
     sprint(print_shell_escaped_winsomely, args..., sizehint=(sum(length, args)) + 3*length(args))
+
+function print_shell_escaped_CMDly(io::IO, arg::AbstractString)
+    any(c -> c in ('\r', '\n'), arg) && throw(ArgumentError("Encountered unsupported character by CMD."))
+    # include " so to avoid toggling behavior of ^
+    arg = replace(arg, r"[%!^\"<>&|]" => s"^\0")
```
Comment (vtjnash) [inline base/shell.jl]:
```suggestion
    arg = replace(arg, r"[%)!^\"<>&|]" => s"^\0")
```
Upon closer inspection of https://stackoverflow.com/questions/4094699/how-does-the-windows-command-interpreter-cmd-exe-parse-scripts/4095133#4095133, I realized that `)` is a metacharacter even though `(` is not. (or we can include both here for symmetry)

**Comment by musm** [comment]
I'm looking back at this PR and I really dread to revisit this PR 😆😆.

PowerShell's handling of external arguments  is so borked that the only reasonable solution is to use the `iep` script in https://github.com/PowerShell/PowerShell/issues/1995#issuecomment-562334606 and  for newer Powershell versions  wait for https://github.com/PowerShell/PowerShell/issues/13068 to be merged

**Comment by StefanKarpinski** [comment]
Counter-proposal: do the same thing on Windows as we do everywhere else, i.e. just exec the cmd that's entered, but ship with busybox tools on Windows and put the path to the busybox tools in the `PATH` before evaluating shell mode commands. I feel like that will make many examples work on Windows in the same way they do on UNIX and doesn't require radically changing the way this works on only one platform.

Also: It would be great (and has been on my todo list for so long) to implement `|` and other POSIX shell features in `Cmd` syntax, in which case the shell mode would automatically get completely cross-platform support for a lot of shell-like features. Combine that with having the busybox tools on Windows and you might not even notice that you're not running on UNIX!

**Comment by stevengj** [comment]
> ship with busybox tools on Windows and put the path to the busybox tools in the PATH before evaluating shell mode commands.

Or just put the path to busybox tools in the PATH for all Cmd evaluations?

**Comment by musm** [comment]
Closing as stale. I'll look at opening a new PR when I find a more satisfactory solution. Shipping busybox is reverting back to something that we removed and suboptimal (even though what we have currently is worse).  Powershell folks have been trying to work around some issues with calling external commands. We updated some the `cmd` escaping rules, so I think we can update to that in another PR.  We might have to limit the powershell option to newer powershell versions and perhaps use bash through WSL if that's available.
